### PR TITLE
[popover] Don't throw when popover/dialog is in requested state

### PR DESCRIPTION
**Pull request by rwlbuis**
#### 1c9711b76a350d4ae9959ddacceb6f7ccc177d22
<pre>
[popover] Don&apos;t throw when popover/dialog is in requested state
<a href="https://bugs.webkit.org/show_bug.cgi?id=255879">https://bugs.webkit.org/show_bug.cgi?id=255879</a>

Reviewed by Tim Nguyen.

Implement the spec changes from:
<a href="https://github.com/whatwg/html/pull/9142">https://github.com/whatwg/html/pull/9142</a>

The dialog related changes are tested by dialog-no-throw-requested-state.html.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-dialog-element/dialog-no-throw-requested-state-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-dialog-element/dialog-no-throw-requested-state.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-attribute-basic.html:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss.html:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-move-documents.html:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/resources/popover-utils.js:
(async sendTab):
(async sendEscape):
(async sendEnter):
(assertPopoverVisibility):
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::show):
(WebCore::HTMLDialogElement::showModal):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::checkPopoverValidity):
(WebCore::HTMLElement::showPopover):
(WebCore::HTMLElement::hidePopoverInternal):

Canonical link: <a href="https://commits.webkit.org/263957@main">https://commits.webkit.org/263957@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/da49f696f134039835bfccde1960344a03c398d0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6217 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6409 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6588 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7785 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6549 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6216 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6629 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6359 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9426 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6327 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6342 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5626 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7856 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3816 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5605 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/13499 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5681 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5685 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/7927 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6164 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/5040 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5577 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/5534 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1475 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9724 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5945 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/html/HTMLDialogElement.cpp [hunk 1]
```diff
@@ -53,8 +53,11 @@ HTMLDialogElement::HTMLDialogElement(const QualifiedName& tagName, Document& doc
 ExceptionOr<void> HTMLDialogElement::show()
 {
     // If the element already has an open attribute, then return.
-    if (isOpen())
-        return { };
+    if (isOpen()) {
+        if (!isModal())
+            return { };
+        return Exception { InvalidStateError, "Element is already open."_s };
```
Comment (nt1m) [inline Source/WebCore/html/HTMLDialogElement.cpp]: `Cannot call show() on an open modal dialog.` ?

File: Source/WebCore/html/HTMLDialogElement.cpp [hunk 2]
```diff
@@ -72,8 +75,11 @@ ExceptionOr<void> HTMLDialogElement::show()
 ExceptionOr<void> HTMLDialogElement::showModal()
 {
     // If subject already has an open attribute, then throw an "InvalidStateError" DOMException.
-    if (isOpen())
+    if (isOpen()) {
+        if (isModal())
+            return { };
         return Exception { InvalidStateError, "Element is already open."_s };
```
Comment (nt1m) [inline Source/WebCore/html/HTMLDialogElement.cpp]: `Cannot call showModal() on an open non-modal dialog.` ?

File: Source/WebCore/html/HTMLElement.cpp [hunk 3]
```diff
@@ -1239,19 +1239,19 @@ ExceptionOr<Ref<ElementInternals>> HTMLElement::attachInternals()
...
+    if (element.popoverData()->visibilityState() != expectedState)
+        return false;
 
     if (expectedDocument && element.document() != *expectedDocument)
         return Exception { InvalidStateError, "Invalid when the document changes while showing or hiding a popover element"_s };
 
-    if (element.popoverData()->visibilityState() != expectedState)
-        return Exception { InvalidStateError, "Element has unexpected visibility state"_s };
+    if (!element.isConnected())
+        return Exception { InvalidStateError, "Element is not connected"_s };
```
Comment (nt1m) [inline Source/WebCore/html/HTMLElement.cpp]: Can you move the connected check before the document check to match the spec?

File: Source/WebCore/html/HTMLElement.cpp [hunk 4]
```diff
@@ -1399,14 +1405,18 @@ ExceptionOr<void> HTMLElement::showPopover()
...
+        if (!check.returnValue())
+            return { };
     }
 
     bool shouldRestoreFocus = !document->topmostAutoPopover();
     RefPtr previouslyFocusedElement = document->focusedElement();
 
-    addToTopLayer();
+    if (!isInTopLayer())
+        addToTopLayer();
```
Comment (nt1m) [inline Source/WebCore/html/HTMLElement.cpp]: Why is this change needed? I don't see it in the spec.

**Review by nt1m — APPROVED** [review]

File: Source/WebCore/html/HTMLDialogElement.cpp (hunk 1 above)
Comment (rwlbuis) [reply Source/WebCore/html/HTMLDialogElement.cpp]: Oops! Fixed.

File: Source/WebCore/html/HTMLDialogElement.cpp (hunk 2 above)
Comment (rwlbuis) [reply Source/WebCore/html/HTMLDialogElement.cpp]: Fixed.

File: Source/WebCore/html/HTMLElement.cpp (hunk 3 above)
Comment (rwlbuis) [reply Source/WebCore/html/HTMLElement.cpp]: Good catch, fixed.

File: Source/WebCore/html/HTMLElement.cpp (hunk 4 above)
Comment (rwlbuis) [reply Source/WebCore/html/HTMLElement.cpp]: Right, I was trying multiple things at this time, but for this patch this change is not needed, removed.
